accepted/future-releases/class-modifiers/feature-specification.md

```diff
@@ -530,62 +576,142 @@ mixinModifier     ::= 'sealed' | 'base' | 'interface' | 'final'
+again be marked `base`, `final`, or `sealed` to avoid it exposing an
+implementable interface.*
+
+Further, while you can ignore some restrictions on declarations within the same
+library, you can't use that to ignore restrictions inherited from other
+libraries.
 
-    A _trivial generative constructor_ is a non-redirecting generative constructor 
-    which has
+We say a class or mixin declaration `D` *can't be implemented locally* if it
+extends, mixes in, implements, or has as on type any declaration `S` where:
 
-    *   an empty parameter list,
-    *   no initializer list (no `: ...`),  
-    *   no constructor body (only `;`),
-    *   and is not marker `external`.
+*   `S` is from another library than `D`, and `S` has the modifier `base`,
+    `final` or `sealed`, or
 
-    Any other generative constructor is non-trivial.
-    A trivial generative constructor may be `const` and may be a named constructor.
+*   `S` is from the same library as `D`, and `S` can't be implemented locally.
 
-    *Declaring a trivial generative constructor allows the class to be used as
-    both a mixin and as a superclass, even if it also declares other factory constructors
-    which suppress the default constructor.*
+Otherwise, `D` can be implemented locally. It is a compile-time error if:
 
-*   Mix in a class not marked `mixin` which has a superclass other than `Object`.
+*   A class or mixin declaration `D` can't be implemented locally, and `D` is
+    not marked `base`, `final` or `sealed`.
 
-*   Mix in a class not marked `mixin` declared in a library with a language version including
-    this feature, if the mixin application is not in the same library, 
-    or if the class has any non-trivial generative constructor.
+*   A class or mixin declaration `D` implements the interface of a class or
+    mixin declaration `S`, declared in the same library as `D`, and `S` can't be
+    implemented locally.
 
-*   Mix in a class not marked `mixin` from a library with a language version older than
-    the version this feature ships in, if the class declares any generative constructor.
+### Mixin restrictions
 
-A typedef can't be used to subvert these restrictions. When extending,
-implementing, or mixing in a typedef, we look at the library where class or
-mixin the typedef resolves to is defined to determine if the behavior is
-allowed. *Note that the library where the _typedef_ is defined does not come
-into play. Typedefs cannot be marked with any of the new modifiers.*
+There are a few changes around mixins to support `mixin class` and disallow
+using normal `class` declarations as mixins while dealing with language
+versioning and backwards compatibility.
+
+Currently, a class may only be used as a mixin if it has a default constructor.
+This prevents the class from defining a `const` constructor or any factory
+constructors. We loosen this somewhat. Define a *trivial generative constructor*
```

Comment:
So to confirm:
```
mixin class M {
  ...
  M(); // 1. This is OK
  const M(); // 2. This is OK
  M(this.x); // 3. This is an error
  M() { ... } // 4. This is an error
  M(): super(); // 5. This is OK
  factory M = _M; // 6. for some _M constructor, this is an error.
}
```

Reply:
`factory M` is OK, it's not a generative constructor. The restriction is only on generative constructors, which must all be trivial.

Reply:
Added examples.